File: 456. 132 Pattern/Java/Solution.java
public class Solution {
    public boolean find132pattern(int[] nums) {
        class Pair {
            int min;
            int max;

            public Pair(int min, int max) {
                this.min = min;
                this.max = max;
            }
        }
        if (nums.length < 3) {
            return false;
        }
        Stack<Pair> stack = new Stack<>();
        for (int num : nums) {
            if (stack.empty() || stack.peek().min > num) {
                stack.push(new Pair(num, num));
            } else if (stack.peek().min < num) {
                if (stack.peek().max > num) {
                    return true;
                }
                int theMin = stack.peek().min;
                stack.pop();
                while (!stack.empty() && stack.peek().max <= num) {
                    stack.pop();
                }
                if (!stack.empty() && stack.peek().min < num) {
                    return true;
                }
                stack.push(new Pair(theMin, num));
            }
        }
        return false;
    }
}

File: 456. 132 Pattern/C++/Solution.cpp
class Solution {
public:
    bool find132pattern(vector<int> &nums) {
        class Pair {
        public:
            int min;
            int max;

            Pair(int min, int max) {
                this->min = min;
                this->max = max;
            }
        };
        if (nums.size() < 3) {
            return false;
        }
        stack<Pair> sta;
        for (int &num : nums) {
            if (sta.empty() || sta.top().min > num) {
                sta.push(Pair(num, num));
            } else if (sta.top().min < num) {
                if (sta.top().max > num) {
                    return true;
                }
                int theMin = sta.top().min;
                sta.pop();
                while (!sta.empty() && sta.top().max <= num) {
                    sta.pop();
                }
                if (!sta.empty() && sta.top().min < num) {
                    return true;
                }
                sta.push(Pair(theMin, num));
            }
        }
        return false;
    }
};